### ISSUE TITLE
Vosk model training process

### ISSUE THREAD

**Issue by IceFloe** [alphacep/vosk-api#182]
Hi! @nshmyrev  

I am trying to train model through mini_librispeech recipe. The main difference that my model should be small, because I am trying to just recognize two words - YES and NO. I am using google commands dataset for this. 

Basically all is good up to tri3b(nosp) step, there I have small model size (5mb for final.mdl), small graph (HCLG - 31kb), nice WER and SER.
decode_nosp_tglarge_dev_clean_2 - wer_17_1.0 - WER 2.31, SER - 2.31
decode_nosp_tgmed_dev_clean_2 - wer_17_1.0 - WER 2.31, SER - 2.31
decode_nosp_tgsmall_dev_clean_2 - wer_17_1.0 - WER 2.31, SER - 2.31
decode_nosp_tgsmall_dev_clean_2.si - wer_17_1.0 - WER 2.69, SER - 2.69
decode_tglarge_dev_clean_2 - wer_17_1.0 - WER 44.23, SER - 42.31
decode_tgmed_dev_clean_2 - wer_17_1.0 - WER 23.08, SER - 21.92
decode_tgsmall_dev_clean_2 - wer_17_1.0 - WER 20.77, SER - 20.00
decode_tgsmall_dev_clean_2.si - wer_17_1.0 - WER 25.38, SER - 24.23

But when I am running last chain script local/chain/tuning/run_tdnn_1j.sh I am receiving bad WER(935.77) and SER(100), vector size 8mb, model size 8mb.
I did some adjustments there and most probably this why it is wrong, but my goal is to have small size model and confident one. 

This changes are:
Using gpu - false, did this on a virtual machine but will install dual boot because I want to use my 2070 gpu.
chunk_width changed to 7,5,8
set for steps/online/nnet2/train_ivector_extractor dim = 30, like you suggested
In neural net config
set input dim=30 name=ivector

  relu-batchnorm-layer name=tdnn1 $tdnn_opts dim=512 input=input2
  tdnnf-layer name=tdnnf2 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=1
  tdnnf-layer name=tdnnf3 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=1
  tdnnf-layer name=tdnnf4 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=1
  tdnnf-layer name=tdnnf5 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=0
  tdnnf-layer name=tdnnf6 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=3
  tdnnf-layer name=tdnnf7 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=3
  tdnnf-layer name=tdnnf8 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=3
  tdnnf-layer name=tdnnf9 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=3
  tdnnf-layer name=tdnnf10 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=3
  tdnnf-layer name=tdnnf11 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=3
  tdnnf-layer name=tdnnf12 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=3
  tdnnf-layer name=tdnnf13 $tdnnf_opts dim=512 bottleneck-dim=64 time-stride=3
  linear-component name=prefinal-l dim=144 $linear_opts
  
  prefinal-layer name=prefinal-chain input=prefinal-l $prefinal_opts small-dim=144 big-dim=512
  output-layer name=output include-log-softmax=false dim=$num_targets $output_opts
  
  prefinal-layer name=prefinal-xent input=prefinal-l $prefinal_opts small-dim=144 big-dim=512
  output-layer name=output-xent dim=$num_targets learning-rate-factor=$learning_rate_factor $output_opts

All other params are the same like in original config.
And the last one
   --trainer.optimization.num-jobs-initial=1 \
   --trainer.optimization.num-jobs-final=3

I am newbie in this training stuff and 100% did some mistakes. Any help would be nice! Thank you! 

The second problem is I am trying to load this model in the Android app and getting E/VoskAPI: KaldiRecognizer():src/kaldi_recognizer.cc:84) Can't create decoding graph. I did not do lookahead script.
Should I run it?

My folder structure is 
word_boundary.int
words.txt
phones.txt
mfcc.conf
HCLG.fst
final.mdl
disambig_tid.int
ivector/splice.conf
ivector/online_cmvn.conf
ivector/global_cmvn.stats
ivector/final.mat
ivector/final.ie
ivector/final.dubm

**Comment by nshmyrev** [comment]
I would say things are not quite reasonable here indeed. What is the final goal of this project?

**Comment by IceFloe** [comment]
I want to have a small model for offline speech recognition on Android. This model will recognize only two words YES or NO. Basically I have found params for chain-nnet3 network which will produce WER around 2.5%, but as I said I am noob in this task and trying to achieve this with "poke" method. 
I don't understand what kind of models can I provide for VOSK.
As well how to have model with low size, I did have course in university far ago about neural networks, so my last thing was to reduce dim for vector to 25 and reduce amount of layers for chain-nnet3 model this will definitely reduce it size. 

You don't have any step by step tutorial how to train model for vosk, that's why it hard to understand what to do.
If I achive my goal I will for sure write tutorial, because what you do is really great.

Like I said any advice from you will help.

**Comment by nshmyrev** [comment]
> If I achive my goal I will for sure write tutorial, because what you do is really great.

Ok, great, let us know if you have one!

**Comment by IceFloe** [comment]
Is vosk is specific for chain-nnet3 model or I could try tri3b model as well, have you tried to use it?
Because in your description of model structure here https://alphacephei.com/vosk/models#training-your-own-model you are saying that ivector folder is optional.

**Comment by nshmyrev** [comment]
> The training is pretty standard - you need tdnn nnet3 model with ivectors.

**Comment by usamaehsan** [comment]
@IceFloe have you writen any step by step tutorial?
I am trying to create ASR  of urdu 
I have data but dont know how to train it

**Comment by IceFloe** [comment]
@usamaehsan I will, but later

**Comment by IceFloe** [comment]
I have tried and found nice coefficients for training, but the biggest problem that kaldi cant work with small vocabulary, it need  to know all possible words if not kaldi will recognize and find the closest word to unknown word.

**Comment by mirfatif** [comment]
@usamaehsan did you ever work on Urdu model?

**Comment by mirfatif** [comment]
@nshmyrev I have a set of less than 10 single words which I need to do STT from English and Urdu on Raspberry Pi and Android. Is there a better way to do this (in 2023, which should also reduce the runtime load on RAM/CPU) other than training a model with huge vocabulary set?

**Comment by nshmyrev** [comment]
> Is there a better way to do this (in 2023, which should also reduce the runtime load on RAM/CPU) other than training a model with huge vocabulary set?

No, you want to model speakers properly, for that you need a large database. It doesn't have to be huge.
